src/consumers.c

```diff
@@ -208,6 +198,123 @@ static void print_metric_meta(FILE* f, metric_metadata* meta) {
+    collection->length = new_length;
+}
+
+/**
+ * Removes item from duration collection
+ * @arg collection - Target collection
+ * @arg value - Value to be removed, assuming primitive type
+ * @return 1 on success
+ */
+/*
+static int remove_bduration_item(bduration_collection* collection, double value) {
+    if (collection == NULL || collection->length == 0 || collection->values == NULL) {
+        return 0;
+    }
+    int removed = 0;
+    long int i;
+    for (i = 0; i < collection->length; i++) {
+        if (removed) {
+            collection->values[i - 1] = collection->values[i];
+        } else {
+            if (*(collection->values[i]) == value) {
+                free(collection->values[i]);
+                removed = 1;
+            }
+        }
+    }
+    if (!removed) {
+        return 0;
+    }
+    collection = realloc(collection, sizeof(double*) * collection->length - 1);
+    ALLOC_CHECK("Unable to resize bduration collection.");
+    collection->length -= 1;
+    return 1;
+}
+*/
+
+static int bduration_values_comparator(const void* x, const void* y) {
+    int res = **(double**)x > **(double**)y;
+    return res;
+}
+
+/**
+ * Gets duration values meta data from given collection, as a sideeffect it sorts the values
+ * @arg collection - Target collection
+ * @arg out - Placeholder for data population
+ * @return 1 on success
+ */
+static int get_bduration_values_meta(bduration_collection* collection, duration_values_meta* out) {
+    if (collection == NULL || collection->length == 0 || collection->values == NULL) {
+        return 0;
+    }
+    qsort(collection->values, collection->length, sizeof(double*), bduration_values_comparator);
+    double accumulator = 0;
+    double min;
+    double max;
+    long int i;
+    for (i = 0; i < collection->length; i++) {
+        double current = *(collection->values[i]);
+        if (i == 0) {
+            min = current;
+            max = current;            
+        }
+        if (current > max) {
+            max = current;
+        }
+        if (current < min) {
+            min = current;
+        }
+        accumulator += current;
+    }
+    out->min = min;
+    out->max = max;
+    out->median = *(collection->values[(int)ceil((collection->length / 2.0) - 1)]);
+    out->average = accumulator / collection->length;
+    out->percentile90 = *(collection->values[((int)round((90.0 / 100.0) * (double)collection->length)) - 1]);
+    out->percentile95 = *(collection->values[((int)round((95.0 / 100.0) * (double)collection->length)) - 1]);
+    out->percentile99 = *(collection->values[((int)round((99.0 / 100.0) * (double)collection->length)) - 1]);
+    out->count = collection->length;
+    accumulator = 0;
+    for (i = 0; i < collection->length; i++) {
+        double x = *(collection->values[i]) - out->average;
+        accumulator += x * x; 
+    }
+    out->std_deviation = sqrt(accumulator / out->count);
+    return 1;
+}
```

Comment:
This function will be called from a different (PCP lib loop) thread while values in the collection can be modified from the other thread (consumer). This needs to be guarded, however the current code is a bit hard to understand - collection is just a parameter of the function and I am not sure what set of functions should be responsible for that data.

It's a good practice to structure C programs in a semi-object oriented pattern where there are data structure(s) at the top of the file and then functions responsible for manipulation. Then it's easier to solve problems like multi threading - it's more clear where to introduce locks. There are libraries which encourages this, for example glibc (GObject) or languages similar to C like Go. I'd suggest to be close to this as possible.

In this case (`consumers.c`) I am not really sure about ownership and responsibilities. Honestly, I'd expect:

* consumer(s).c - defining consumers API - set of functions and base structures common to all consumers
* consumer-exact.c - data structures (collections) and implementation for exact consumer
* consumer-hdr.c - ditto for HDR

Then the thread-safety mechanism should belong to `consumer-exact.c` only in this case.

Reply:
Code structure updated in 423d74f. Will add thread-safety to exact duration metric handling soon.

Reply:
Why should thread-safety be only in exact duration consumer? I looked at the definitions of hdr_histogram and there are no locks. I should probably add it there as well?

Reply:
I was thinking a lot about how to make this thread-safe (also discussing with some people on the internet) and wouldn't it be better if the aggregation thread took over all ownership of metric data so that no thread-races would be possible while acessing it? With such solution I would just create another channel between PCP-thread and AGREGATION-thread via which the PCP would send messages specifying which data it would want to access (or when I should send over all collected data). The data could be then fully managed by AGGREGATION-thread and automatically deleted when after being requested from PCP-thread.

Reply:
I did not want to mislead you - thread safety is indeed required for all consumers, including HDR one. Anyway.

Using channels is indeed a good design, you actually have them available in your codebase and it's a good idea to build as much as you can on them. So I like the idea of separation of concerns. However there is one particular problem and that is data flow - packets, lines and parsed structure flow one way while PCP client thread callback is called regularly (every second usually). You need to find mechanism to tell the aggregation component to send the data (via channel). AFAIK channel is only one-way communication structure, you can probably create a new one. Or if it is not, then you could use it.

Overall I support the idea of using channels as much as possible, they appear to be very safe programming pattern in Go and you can apply the same here.

Reply:
> However there is one particular problem and that is data flow - packets, lines and parsed structure flow one way while PCP client thread callback is called regularly (every second usually).

The while loop in consumer thread uses chan_select which is actually non-blocking, so I can just append another such select which will listen to PCP-thread, so this is non-issue. If needed will make two of them for full-duplex communication.

I will create such mechanism then - what kind of messages should I expect from PCP thread? Something akin to "GET <metricname> <metrictype>" (returns value of such metric) or "GET ALL RECORDS" (returns name + value of all recorded metrics)? First option may take longer to execute as the loop will continue to cycle after each command (should there be many) and as such may also throttle itself recording values from parser thread.

Reply:
Please read the PCP Developer Handbook in order to understand how PMDA agents are designed, I remember that metrics are queried one by one, so the interface should probably be just "get me aggregated value for metric XYZ". I do not know if labels are returned all at once or individually too - the last time I checked PCP did not support them.